psl-core/src/main/java/org/linqs/psl/reasoner/dcd/DCDReasoner.java

```diff
@@ -59,63 +60,79 @@ public double optimize(TermStore baseTermStore) {
 
         termStore.initForOptimization();
 
-        // This must be called after the term store has to correct variable capacity.
-        // A reallocation can cause this array to become out-of-date.
-        float[] variableValues = termStore.getVariableValues();
-
-        double objective = -1.0;
+        long termCount = 0;
+        double objective = Double.POSITIVE_INFINITY;
         double oldObjective = Double.POSITIVE_INFINITY;
+        float[] oldVariableValues = null;
 
-        if (printInitialObj && log.isTraceEnabled()) {
-            objective = computeObjective(termStore, variableValues);
-            log.trace("Iteration {} -- Objective: {}, Iteration Time: {}, Total Optimiztion Time: {}", 0, objective, 0, 0);
+        if (log.isTraceEnabled()) {
+            objective = computeObjective(termStore);
+            log.trace("Iteration {} -- Objective: {}, Normalized Objective: {}, Mean Movement: {}, Iteration Time: {}, Total Optimization Time: {}",
+                    0, objective, objective / termStore.size(), 0, 0, 0);
         }
 
         int iteration = 1;
         long totalTime = 0;
         while (true) {
             long start = System.currentTimeMillis();
 
+            termCount = 0;
+            objective = 0;
             for (DCDObjectiveTerm term : termStore) {
```

Comment:
Did you test DCD convergence? i remember there was some issue and it wouldnt converge smoothly.

Reply:
DCD seems to be converging but to suboptimal points (based on its normalized objective compared to SGD).
This can be observed in the [google sheet](https://drive.google.com/drive/folders/1MPTfYrqD_d-MspsUs6iu8CqAYPXz0i1s?usp=sharing).

Reply:
Thats strange. You should let it run for long enough to see what happens.

Reply:
Also did you try on other datasets??